### ISSUE TITLE
Flatpak/Flathub distribution

### ISSUE THREAD

**Issue by zamdzhiev** [tikzit/tikzit#25]
It looks like the portable binary and the Debian package work pretty well and these are probably the best options for most people.

Another option is to distribute this in Flatpak format on Flathub. I have already written the necessary manifest for it and TikZiT builds and runs fine as a flatpak app. Here's the manifest (a JSON file):

```
{
    "app-id": "io.github.tikzit",

    "runtime": "org.kde.Platform",
    "runtime-version": "5.11",

    "sdk": "org.kde.Sdk",
    "command": "tikzit",
    "finish-args": ["--socket=x11",
                    "--socket=wayland",
                    "--share=ipc",
                    "--device=dri",
                    "--filesystem=host"
                   ],
    "modules": [
        {
            "name": "tikzit",
            "buildsystem": "simple",
            "build-commands": [
                "qmake",
                "make",
                "install -D tikzit /app/bin/tikzit"
            ],
            "sources": [
                {
                    "type": "git",
                    "url": "https://github.com/tikzit/tikzit.git"
                }
            ]
        }
    ]
}
```

Then flatpak-builder would clone the repository and build it as a flatpak app (you can also specify git tags to checkout).

Flatpak apps are run sandboxed together with all libraries they need and are easy to install from Flathub (basically an app store for all Linux distributions). They are supposed to run on any Linux distribution. One of the downsides is that flatpak will pull in a lot of dependencies if they are not already present (in this case, that's 300+MB). But if they are, then it won't. There are other pros and cons that you can read about.

In any case, this is an extra distribution channel and might be useful for some people. Also, TikZiT becomes more easily discoverable if it is available on Flathub.

If you want, I can submit it to Flahub and maintain it there.

**Comment by akissinger** [comment]
Yes, that would be great. Let me know if I should do anything on the tikzit side.

**Comment by JackDeSomeTrades** [comment]
@zamdzhiev and @akissinger  was the flatpak app put on flathub? I'm trying to get TikZiT to compile on Ubuntu 16.04 and it is proving to be harder than expected. A flatpak app would be great. Thanks.

**Comment by zamdzhiev** [comment]
Unfortunately, no. I cannot make a flatpak build of the newest version with PDF preview support. I can only build the 2.0 zero version as a flatpak app. I don't want to put an old flatpak on Flathub, so I first need to figure out how to produce a working flatpak of the newest versions.

Do you have experience with flatpak packaging? I really can't figure out why the build is failing on my Debian laptop. I will soon put a newer version of Debian on my dekstop (with newer flatpak-builder) and try to make a new flatpak build there. I will let you know how it goes.

**Comment by JackDeSomeTrades** [comment]
I have the same problem. I used your manifest to try and build the newest github version and the build seems to be failing. I am not very experienced in using the flatpak-builder so that could be a reason. 

The build is failing on my Ubuntu laptop because it can't find poppler libraries even though I have them installed. Maybe there needs to be something more in the manifest? This is my current manifest that's failing. 


`
{
    "app-id": "io.github.tikzit",

    "runtime": "org.kde.Platform",
    "runtime-version": "5.11",

    "sdk": "org.kde.Sdk",
    "command": "tikzit",
    "finish-args": ["--socket=x11",
                    "--socket=wayland",
                    "--share=ipc",
                    "--device=dri",
                    "--filesystem=host"
                   ],
    "modules": [
    	{
            "name": "popplerdata",
            "no-autogen": true,
            "make-install-args": [
                "prefix=/app"
            ],
            "sources": [
                {
                    "type": "archive",
                    "url": "http://poppler.freedesktop.org/poppler-data-0.4.9.tar.gz",
                    "sha256": "1f9c7e7de9ecd0db6ab287349e31bf815ca108a5a175cf906a90163bdbe32012"
                }
            ]
        },
        {
            "name": "poppler",
            "buildsystem": "cmake-ninja",
            "config-opts": [
                "-DCMAKE_INSTALL_LIBDIR=/app/lib",
                "-DCMAKE_INSTALL_INCLUDEDIR=/app/include",
                "-DENABLE_LIBOPENJPEG=none"
            ],
            "cleanup": [
                "/bin"
            ],
            "sources": [
                {
                    "type": "archive",
                    "url": "https://poppler.freedesktop.org/poppler-0.72.0.tar.xz",
                    "sha256": "c1747eb8f26e9e753c4001ed951db2896edc1021b6d0f547a0bd2a27c30ada51"
                }
            ]
        },
        {
            "name": "tikzit",
            "buildsystem": "simple",
            "build-commands": [
                "qmake",
                "make",
                "install -D tikzit /app/bin/tikzit"
            ],
            "sources": [
                {
                    "type": "git",
                    "url": "https://github.com/tikzit/tikzit.git"
                }
            ]
        }
    ]}`

**Comment by akissinger** [comment]
Poppler needs to be built with the Qt bindings, which I believe are not enabled by default. I think you need to pass a flag to `./configure` if you are building from source. On Ubuntu, you need to apt-get both `libpoppler-dev` and `libpoppler-qt5-dev`.

**Comment by zamdzhiev** [comment]
Yes, to build it as a flatpak app you need to make a module for poppler. However, poppler itself depends on some other stuff (openjpeg). The following build script almost succeeds for me, but fails at the very end when it tries to link. I can compile all libraries (including libpoppler-qt5) and I can see them in the build directory, but it fails to link for some reason. I don't know why.

@JackDeSomeTrades can you please try and see if my build script works on your machine? If it doesn't, maybe try to add the openjpeg module to your build-script?

``
{
    "app-id": "io.github.tikzit",
    "runtime": "org.kde.Platform",
    "runtime-version": "5.11",
    "sdk": "org.kde.Sdk",
    "command": "tikzit",
    "finish-args": ["--socket=x11",
                    "--socket=wayland",
                    "--share=ipc",
                    "--device=dri",
                    "--filesystem=host"
                   ],
    "modules": [
        {
            "name": "openjpeg",
            "buildsystem": "cmake",
            "sources": [
                {
                    "type": "git",
                    "url": "https://github.com/uclouvain/openjpeg.git"
                }
            ]
        },
        {
            "name": "poppler",
            "buildsystem": "cmake",
            "sources": [
                {
                    "type": "git",
                    "url": "https://anongit.freedesktop.org/git/poppler/poppler.git"
                }
            ]
        },
        {
            "name": "tikzit",
            "buildsystem": "simple",
            "build-commands": [
                "qmake -r",
                "make",
                "install -D tikzit /app/bin/tikzit"
            ],
            "sources": [
                {
                    "type": "git",
                    "url": "https://github.com/tikzit/tikzit.git"
                }
            ]
        }
    ]
}

**Comment by JackDeSomeTrades** [comment]
@zamdzhiev I tried your manifest and I am getting the same error as you did. 

`cannot find -lpoppler-qt5`
`collect2: error: ld returned 1 exit status`  
`make: *** [Makefile:469: tikzit] Error 1`

**Comment by zamdzhiev** [comment]
What version of flatpak-builder are you using? I am on flatpak-builder 1.0.1.

Do you think this is a bug with flatpak-builder or do you think we are doing something wrong? I really can't tell.

Do you know of a good place where we can ask some flatpak wizards for advice? There is always the possibility of opening an issues on the flatpak github page, but this might annoy the devs if it is not actually a bug and we are missing something.

**Comment by JackDeSomeTrades** [comment]
I'm using the same version of flatpak-builder. 

I tried to use your manifest with mine as well and see if it works, but no luck. At this point, I can't tell either. I think it has to do with the build structure of flatpack-builder and where/how it sets the LD_LIBRARY_PATH. I've given up. 

Asking the devs might be a good idea.

**Comment by zamdzhiev** [comment]
I tried building it with the latest flatpak-builder, but no luck. Same error as before.

Also, I updated the manifest, so that it refers to stable version numbers only (provided below).

I opened up an issue on the flatpak-builder github page. Perhaps the developer might give us some useful information.

https://github.com/flatpak/flatpak-builder/issues/265

```
{
    "app-id": "io.github.tikzit",
    "runtime": "org.kde.Platform",
    "runtime-version": "5.11",
    "sdk": "org.kde.Sdk",
    "command": "tikzit",
    "finish-args": ["--socket=x11",
                    "--socket=wayland",
                    "--share=ipc",
                    "--device=dri",
                    "--filesystem=host"
                   ],
    "modules": [
        {
            "name": "openjpeg",
            "buildsystem": "cmake",
            "sources": [
                {
                    "type": "archive",
                    "url": "https://github.com/uclouvain/openjpeg/archive/v2.3.0.tar.gz",
                    "sha256": "3dc787c1bb6023ba846c2a0d9b1f6e179f1cd255172bde9eb75b01f1e6c7d71a"
                }
            ]
        },
        {
            "name": "poppler",
            "buildsystem": "cmake",
            "sources": [
                {
                    "type": "archive",
                    "url": "https://poppler.freedesktop.org/poppler-0.72.0.tar.xz",
                    "sha256": "c1747eb8f26e9e753c4001ed951db2896edc1021b6d0f547a0bd2a27c30ada51"
                }
            ]
        },
        {
            "name": "tikzit",
            "buildsystem": "simple",
            "build-commands": [
                "qmake -r",
                "make",
                "install -D tikzit /app/bin/tikzit"
            ],
            "sources": [
                {
                    "type": "archive",
                    "url": "https://github.com/tikzit/tikzit/archive/v2.1.3.tar.gz",
                    "sha256": "01638ea9e2f883f33f282f681bddc21410ea6b5c7709647bd6eb61dca0d7c1c7"
                }
            ]
        }
    ]
}
```

**Comment by akissinger** [comment]
I just updated CMakelists.txt, so in principle, tikzit should be cmake-able, provided <a href="https://api.kde.org/ecm/manual/ecm.7.html">ECM</a> is available. You could try the "cmake" buildsystem rather than "simple" (though you should make sure cmake actually works first, since I haven't tested it in a while).

**Comment by zamdzhiev** [comment]
I just tried it, but it doesn't work. I tried compiling from source using cmake and also tried it using flatpak-builder. Both times I get an error

```
-- The CXX compiler identification is GNU 6.4.0
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found BISON: /usr/bin/bison (found version "3.0.4") 
-- Found FLEX: /usr/bin/flex (found version "2.6.0") 
-- Found PkgConfig: /usr/bin/pkg-config (found version "0.29.1") 
-- Found Poppler_Core: /app/lib/libpoppler.so (found version "0.72.0") 
-- Found Poppler_Qt5: /app/lib/libpoppler-qt5.so (found version "0.72.0") 
-- Found Poppler: /app/lib/libpoppler.so;/app/lib/libpoppler-qt5.so (found version "0.72.0") found components:  Core Qt5 
CMake Error at CMakeLists.txt:28 (cmake_policy):
  Policy "CMP0071" is not known to this version of CMake.

```

**Comment by akissinger** [comment]
Try commenting out the line that sets the policy on CMakelists.txt (probably this should be wrapped in a conditional, but I don't know cmake syntax very well....)

**Comment by zamdzhiev** [comment]
I can build with cmake from source, but it needs some changes. If you can incorporate the following CMakeLists.txt (delete commented lines as you wish) in the source tree that would be nice.

I tried to make the tests work as well, but I couldn't, so I just commented out the tests. However, it is necessary to remove the lexer and parser stuff from the sources (because they have to be generated first by flex and bison).

With this cmake file I can generate and build tikzit on my machine and it runs fine.

The good news is that flatpak-builder successfully links and compiles tikzit. The only problem remaining is to have a "make install" target. Since this is not there, it does not know where to put the binary. But it successfully builds everything, so that's good and we are almost done.

So we just have to find some documentation about this and we're done.

```
cmake_minimum_required(VERSION 3.1.0)

project(tikzit LANGUAGES CXX)

#enable_testing()

set(CMAKE_INCLUDE_CURRENT_DIR ON)
set(CMAKE_AUTOMOC ON)
#set(CMAKE_AUTOUIC ON)

find_package(Qt5 REQUIRED COMPONENTS Core Gui Widgets Network)
#find_package(Qt5Test REQUIRED)
find_package(BISON)
find_package(FLEX)

# use extra-cmake-modules to find poppler library
find_package(ECM REQUIRED NO_MODULE)
set(CMAKE_MODULE_PATH ${ECM_FIND_MODULE_DIR})
find_package(Poppler REQUIRED COMPONENTS Core Qt5)

include_directories(
    ${PROJECT_SOURCE_DIR}
    ${PROJECT_SOURCE_DIR}/src
    ${PROJECT_SOURCE_DIR}/src/data
    ${PROJECT_SOURCE_DIR}/src/gui
)

BISON_TARGET(PARSER src/data/tikzparser.y ${CMAKE_CURRENT_BINARY_DIR}/tikzparser.parser.cpp)
FLEX_TARGET(LEXER src/data/tikzlexer.l ${CMAKE_CURRENT_BINARY_DIR}/tikzlexer.lexer.cpp)
ADD_FLEX_BISON_DEPENDENCY(LEXER PARSER)
#cmake_policy(SET CMP0071 NEW) # run AUTOXXX on generated files ('NEW' default behaviour)

set(SOURCES
    src/data/delimitedstringvalidator.cpp
    src/data/edge.cpp
    src/data/graph.cpp
    src/data/graphelementdata.cpp
    src/data/graphelementproperty.cpp
    src/data/node.cpp
    src/data/pdfdocument.cpp
    src/data/style.cpp
    src/data/stylelist.cpp
    src/data/tikzassembler.cpp
    src/data/tikzdocument.cpp
#    src/data/tikzlexer.lexer.cpp
#    src/data/tikzparser.parser.cpp
    src/data/tikzstyles.cpp
    src/gui/commands.cpp
    src/gui/delimitedstringitemdelegate.cpp
    src/gui/edgeitem.cpp
    src/gui/exportdialog.cpp
    src/gui/latexprocess.cpp
    src/gui/mainmenu.cpp
    src/gui/mainwindow.cpp
    src/gui/nodeitem.cpp
    src/gui/preferencedialog.cpp
    src/gui/previewwindow.cpp
    src/gui/propertypalette.cpp
    src/gui/styleeditor.cpp
    src/gui/stylepalette.cpp
    src/gui/tikzscene.cpp
    src/gui/tikzview.cpp
    src/gui/toolpalette.cpp
    src/gui/undocommands.cpp
    src/main.cpp
#    src/test/testmain.cpp
#    src/test/testparser.cpp
#    src/test/testtest.cpp
    #src/test/testtikzoutput.cpp
    src/tikzit.cpp
    src/util.cpp
)

set(HEADERS
    src/data/edge.h
    src/data/graph.h
    src/data/graphelementdata.h
    src/data/graphelementproperty.h
    src/data/node.h
    src/data/pdfdocument.h
    src/data/style.h
    src/data/stylelist.h
    src/data/tikzassembler.h
    src/data/tikzdocument.h
    src/data/tikzparserdefs.h
    src/data/tikzstyles.h
    src/gui/commands.h
    src/gui/delimitedstringitemdelegate.h
    src/gui/edgeitem.h
    src/gui/exportdialog.h
    src/gui/latexprocess.h
    src/gui/mainmenu.h
    src/gui/mainwindow.h
    src/gui/nodeitem.h
    src/gui/preferencedialog.h
    src/gui/previewwindow.h
    src/gui/propertypalette.h
    src/gui/styleeditor.h
    src/gui/stylepalette.h
    src/gui/tikzscene.h
    src/gui/tikzview.h
    src/gui/toolpalette.h
    src/gui/undocommands.h
#    src/test/testparser.h
#    src/test/testtest.h
#    src/test/testtikzoutput.h
    src/tikzit.h
    src/util.h
)

set(FORMS
    src/gui/exportdialog.ui
    src/gui/mainmenu.ui
    src/gui/mainwindow.ui
    src/gui/preferencedialog.ui
    src/gui/previewwindow.ui
    src/gui/propertypalette.ui
    src/gui/styleeditor.ui
    src/gui/stylepalette.ui
)

qt5_wrap_ui(FORM_HEADERS ${FORMS})
qt5_add_resources(QT_RESOURCES tikzit.qrc)



add_executable(tikzit
    ${SOURCES}
    ${HEADERS}
    ${FORM_HEADERS}
    ${FLEX_LEXER_OUTPUTS}
    ${BISON_PARSER_OUTPUTS}
    ${QT_RESOURCES})

#add_test(tikzit tikzit)

target_link_libraries(${PROJECT_NAME}
    Qt5::Core Qt5::Gui Qt5::Widgets Qt5::Network
#    Qt5::Test
    ${Poppler_LIBRARIES})

```

**Comment by zamdzhiev** [comment]
At the bottom you just have to add

`install (TARGETS tikzit DESTINATION bin)`

and it does the trick.

**Comment by zamdzhiev** [comment]
OK, so I can build the flatpak now and I can run it, but the PDF preview doesn't work (from within the flatpak sandbox). The problem is that flatpak applications are run in a sandbox and they are not really meant to execute outside binaries in general.

The sandboxed tikzit won't find pdflatex in the usual way. The way out of this, is to execute a command called `flatpak-spawn` which allows a sanboxed application to break out of its sandbox and interact with the host system.

I modified the tikzit source code (simple edit to latexprocess.cpp) so that tikzit calls `flatpak-spawn --host pdflatex ...` but then unfortunately I run into this flatpak bug: https://github.com/flatpak/flatpak/issues/2418 and I cannot do anything about this. The bug was fixed very recently, but the updated version has not propagated to most distros yet. Not even Debian Unstable has it yet. So, I cannot do any more tests until I receive an update, but I think this might work.

So, even if we actually code tikzit to detect when it is in a sandbox and break out of it, we still have to wait for the version of flatpak that fixes the bug to be delivered to the users, because otherwise there's no chance for PDF preview to work for them. This will be several months at least in my opinion...

That's disappointing, but I don't see what else can be done about this.

**Comment by akissinger** [comment]
Done. A suggestion: rather than changing the tikzit code, can you add a wrapper script for pdflatex in the sandboxed $PATH that calls it via flatpak-spawn and passes all the arguments through?

**Comment by zamdzhiev** [comment]
Yes, I think this might work. Basically, we ship an extra bash script called ``pdflatex`` which we put in the sandboxed ``/usr/bin``, then the sandboxed ``tikzit`` will think that this is pdflatex, but the script will just break out of the sandbox and call the host's actual ``pdflatex`` with the required arguments. I think this can be done using the ``post-install`` option of the manifest of flatpak-builder.

I will test this after I receive an update to flatpak on my desktop, because it clearly can't work before that happens.

**Comment by akissinger** [comment]
I'm thinking of closing this issue. Did anything more happen with this in the past year?
